benchmark/src/pre_benchmarks.cpp

```diff
@@ -0,0 +1,319 @@
+// OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+// OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//==================================================================================
+
+#include "benchmark/benchmark.h"
+
+#include "openfhe.h"
+
+#include "scheme/bgvrns/cryptocontext-bgvrns.h"
+#include "gen-cryptocontext.h"
+
+#include <chrono>
+#include <fstream>
+#include <iostream>
+#include <iterator>
+#include <getopt.h>
+
+
+using namespace std;
+using namespace lbcrypto;
+
+int num_of_hops=1; //number of hops
+usint security_model = 0; //0 - CPA secure PRE, 1 - fixed 20 bits noise, 2 - provable secure HRA noise flooding with BV key switching,
+                          // 3 - provable secure HRA noise flooding with Hybrid key switching, 4 - provably secure HRA without noise flooding
+
+void usage() {
+  std::cout << "-m security model (0 CPA secure PRE, 1 Fixed 20 bits noise, 2 Provable secure HRA)"
+            << "-d number of hops"
+            << std::endl;
+}
+
+
+//Default set to IND-CPA parameters
+//The parameters ring_dimension, relinWindow, qmodulus are chosen for a plaintext modulus of 2 to allow upto 10 hops 
+//without breaking decryption. Changing the plaintext modulus would require updating these other parameters for correctness.
+lbcrypto::CCParams<lbcrypto::CryptoContextBGVRNS> parameters;
+
+int plaintextModulus = 2;
+uint32_t multDepth = 0;
+double sigma = 3.2;		
+SecurityLevel securityLevel = HEStd_128_classic;
+usint ringDimension = 1024;
+usint relinWindow = 3;
+usint dcrtbits = 0;
+
+usint qmodulus = 27;
+usint firstqmod = 27;
+  
+lbcrypto::CryptoContext<lbcrypto::DCRTPoly> cryptoContext;
+// Initialize Key Pair Containers
+lbcrypto::KeyPair<lbcrypto::DCRTPoly> keyPairproducer, keyPairconsumer;
+
+vector<int64_t> vectorOfInts;
+unsigned int nShort=0;
+int ringsize=0;
+
+Plaintext plaintext, plaintextDecproducer, plaintextDecconsumer;
+lbcrypto::Ciphertext<DCRTPoly> ciphertext1, reEncryptedCT1, reEncryptedCT;
+
+lbcrypto::EvalKey<DCRTPoly> reencryptionKey;
+
+void PRE_keygen(benchmark::State &state) {
+    //std::cout << "\nRunning key generation (used for source data)..."
+    //            << std::endl;
+
+    for (auto _ : state) {
+        keyPairproducer = cryptoContext->KeyGen();
+    }
+
+    if (!keyPairproducer.good()) {
+        std::cout << "Key generation failed!" << std::endl;
+        exit(1);
+    }
+}
+void PRE_Encrypt(benchmark::State &state) {
+    //std::cout << "\nRunning encrypt (used for source data)..."
+    //            << std::endl;
+    ringsize = cryptoContext->GetRingDimension();
+    nShort = ringsize;
+    for (size_t i = 0; i < nShort; i++){ //generate a random array of shorts
+        if(plaintextModulus==2) {
+            vectorOfInts.push_back(std::rand() % plaintextModulus);
+        }
+        else {
+            vectorOfInts.push_back((std::rand() % plaintextModulus) - (std::floor(plaintextModulus/2)-1));
```

Comment:
the std::floor is applied to plaintextModulus/2 which might not always be an integer if plaintextModulus is odd.

Reply:
you're wrong. the result is always an integer as plaintextModulus is an integer

Reply:
Ok so you mean it is an integer type in C++? Because sometimes we use 65537 as the plaintext modulus, so 65537/2 is not an integer. I will remove the floor